design-proposals/dedicated-migration-network.md

```diff
@@ -0,0 +1,129 @@
+
+## Motivation
+In order for VMI migrations to succeed, there must be a fair amount of bandwidth available.  
+If for example the memory of a migrating VMI changes faster than the current speed of the main network, the migration will never succeed.  
+This is especially problematic during mass eviction events like platform upgrades.  
+This problem is currently mitigated by throttling the number of concurrent migrations and using long timeouts.
+
+## Goals
+- Allow a cluster admin to setup and dedicate a secondary physical cluster network to migrations.
+- Keep the main network as the default medium for migrations.
+
+## Non Goals
+- QoS. That is an obvious next step after enabling the use of secondary networks. However, this is out of scope for the initial implementation.
+- Advanced routing. The secondary network is expected to assign IPs on a subnet that does not intersect with the subnet of the main network. Advanced/customized routing is out of scope for this proposal.
+- Network configuration. A functional secondary physical network should be provided.
+- Network definition. A valid NetworkAttachmentDefinition should be written by the cluster admin and provided to the KubeVirt CR (Custom Resource).
+- Cluster components. A multi-network CNI like multus should be properly installed, as well as necessary plugins, and any required daemon should be running on each node.
+
+## User story
+As an administrator of the cluster I want all my VMIs to be migrated over a dedicated network to provide a stable connection with a larger bandwidth comparing to the default k8s pod network
+
+## Definition of Users
+This feature is intended for cluster admins who wish to accelerate and/or control migration network usage.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt):
+  - add a CR option and handle it at the virt-handler level
+- [kubevirt/kubevirtci](https://github.com/kubevirt/kubevirtci):
+  - install [whereabouts](https://github.com/k8snetworkplumbingwg/whereabouts) on network test lanes
+  - add a secondary network to network lanes
+
+# Design
+If a dedicated migration network is defined in the KubeVirt CR, migration target virt-handler should communicate their IP of that network instead of the main one.   
+Dedicated migration networks will be exposed to virt-handlers as "migration0". Their presence and assigned IP can be discovered using the [Downward API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api).  
+The source virt-handler will then use that IP to communicate with the target, which will get routed through the secondary network.
+
+## API
+In the KubeVirt CR, the ability to define network roles will be added under spec.configuration.network, like so:
+```
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - LiveMigration
+    network:
+      roles:
+      - networkAttachmentDefinition: my-secondary-network
```

Comment:
> In this case the evolution for this possible additional use-case would be possible without disturbance, but we can still focus on the main use-case. Further the API field naming gives very good hints regarding to what these fields may do.
> 
> If this section is then also placed in the migration section, also a more crisp representation may be possible without loosing meaning:

I understand the advantages here, this is a good argument.
Nevertheless, I think that there are also advantages in centralizing networking aspects under the same "roof".
I will try to outline my view on this with some scenarios.

---

Summarizing the advantages of the current proposal with a use-case beyond "migration":
- Use case (optional extensions):
  - Terminal (view) network: Currently, accessing the terminal (console/vnc) is done through the API server. A potential dedicated network for this task may make sense. The reason is very similar to the migration network, one wants  a dedicated network to work with in order to assure continuation of the service even if others are overloaded or some services are down.
  - It may be useful in the future to let the Kubevirt cluster operator (and not the k8s operator) to limit which networks can be used by the VM/s. One optional way to do this may be to list the "approved" networks. I'm just playing with the idea, to illustrate how this can evolve beyond the single need we have in front of us today.

- When defining a secondary network, there is some networking related knowledge needed from the operator. Operators have sometime domain expertise, like we do in our SIG separation. Today we already have potential network specific parameters that are required, the minimum being the NAD and from there the pod interface name. It is likely that we will need other details to control this interface (through CNI pod annotations), therefore more fields may be needed.
All this sums up with the obvious need to understand and know what a dedicated network means and what are its values. It is challenging IMO for one operator to know it all, and such a domain separation and centralization makes it more maintainable and extendable.

- Continuing the previous point, centralizing network services is a common organization in API/s, at least for other solutions that exist on the market. I guess the main point would sum up to be: How subjects and features should be organized. The approach I favor is per service domain and the other one is feature centric.
There is one nice advantage to all such systems, they cause a generic format for providing the service to whatever feature/sub-system that requires it.

- It is preferable not to need and change the API such that it will cause backward compatibility issues. We saw bad examples of this in k8s when IPv6 was not though of from day one (duplication and backward compatibility with services).

I find the current proposal better structured due to its centralization. Having network parts all over the configuration spec looks messy to me. I do have hard time imagining that this will be the only network option we will add, therefore my opinion on this.

Reply:
> * Terminal (view) network: Currently, accessing the terminal (console/vnc) is done through the API server. A potential dedicated network for this task may make sense. The reason is very similar to the migration network, one wants  a dedicated network to work with in order to assure continuation of the service even if others are overloaded or some services are down.

There are no plans to do that. VNC and console are administrative minimal access points. We btw. also have `port-forward` and `ssh` now. KubeVirt is not a out-of-the-box solution for e.g. remote desktop.

General Traffic shaping for workload ports, no matter if VMs or Pods is in general nice, but I personally don't see this happening in the KubeVirt CR.

> * It may be useful in the future to let the Kubevirt cluster operator (and not the k8s operator) to limit which networks can be used by the VM/s. One optional way to do this may be to list the "approved" networks. I'm just playing with the idea, to illustrate how this can evolve beyond the single need we have in front of us today.

This is something outside of the scope of kubevirt. What people are allowed to create is a matter of RBAC and access control where there is no need to separate between pods and VMs. Sending traffic over a network with VMs should not be more harmful to the cluster than sending it over pods.

Even if so, like above, I see this in the context of general purpose workload traffic shaping which I don't see happening in the KubeVirt CR.

> * When defining a secondary network, there is some networking related knowledge needed from the operator. Operators have sometime domain expertise, like we do in our SIG separation. Today we already have potential network specific parameters that are required, the minimum being the NAD and from there the pod interface name. It is likely that we will need other details to control this interface (through CNI pod annotations), therefore more fields may be needed.

I am not aware of those. Can you please elaborate? We need all kinds of realistic examples which we can get.

>   All this sums up with the obvious need to understand and know what a dedicated network means and what are its values. It is challenging IMO for one operator to know it all, and such a domain separation and centralization makes it more maintainable and extendable.
> * Continuing the previous point, centralizing network services is a common organization in API/s, at least for other solutions that exist on the market. I guess the main point would sum up to be: How subjects and features should be organized. The approach I favor is per service domain and the other one is feature centric.

If we can't create a critical mass of examples which will happen in kubevirt (and we are already a few years in development now and did not see this need emerge), I would continue to favour the feature centric approach and worst-case maintain the deprecation of a simple field which is entirely under our control.


>   There is one nice advantage to all such systems, they cause a generic format for providing the service to whatever feature/sub-system that requires it.

It would only be generic if we can expect that all feature related knobs would not be in that list. That would be unrealistic. For migrations we have a great example where it would actually be split the feature settings.

> * It is preferable not to need and change the API such that it will cause backward compatibility issues. We saw bad examples of this in k8s when IPv6 was not though of from day one (duplication and backward compatibility with services).

Yes this is however only a minimal issue if we would have to deprecate the field. It is pretty much irrelevant. I think you have to take comparable examples. All the code is under our control which relies on this. So it is safe to wait for use-cases to emerge, which they potentially will not, or only after a very long time period.

Reply:
> > * When defining a secondary network, there is some networking related knowledge needed from the operator. Operators have sometime domain expertise, like we do in our SIG separation. Today we already have potential network specific parameters that are required, the minimum being the NAD and from there the pod interface name. It is likely that we will need other details to control this interface (through CNI pod annotations), therefore more fields may be needed.
> 
> I am not aware of those. Can you please elaborate? We need all kinds of realistic examples which we can get.

Here for instance, if it turns out that we need such a section for workload ports to properly integrate within existing networks (e.g. to function at all), independent of traffic shaping, I could imagine for that feature (workload networks) an extra section where a list of additional config hints makes sense.  We can again ask us here if we want then to combine workload network (vm traffic) with administrative networks (e.g. migration network).